### ISSUE TITLE
[Regression] script hangs in 0.2.0, not 0.1.5 when not using context manager

### ISSUE THREAD

**Issue by mdavis-xyz** [pahaz/sshtunnel#201]
I'm finding that my scripts are hanging forever even after I close the tunnel. This happens in versions 0.2.1 and 0.2.0, but not 0.1.5. So I think something has happened to break this.

Note that in my use case I do not want to use a context manager. I'm deploying inside an AWS Lambda function, and I want to cache the connection between Lambda invocations, which means I need to assign the tunnel object to a global variable.
I also have some unit tests with the `unittest` module's `setUp()` and `tearDown()` functions.

The difference between `.start()` and the context manager seems to be:

```
.daemon_forward_servers = True
.daemon_transport = True
```

What's the meaning of these parameters? Why are the defaults different between context manager and a function?

# Steps to reproduce

We'll SSH to `localhost`, so run:

```
ssh key-gen
cat ~/.ssh/id_rsa.pub >> ~/.ssh/authorized_keys
```
Then run this script:

```
from sshtunnel import SSHTunnelForwarder
from time import sleep

server = SSHTunnelForwarder(
    ssh_address_or_host='localhost',
    remote_bind_address=('localhost', 80),
    local_bind_address=('localhost', 8088),
    ssh_username='ec2-user',
    ssh_port=22,
    ssh_pkey='~/.ssh/epic_ssh_unit_test_key.private',
    ssh_config_file=None,
    allow_agent=False,
    set_keepalive=10,
    threaded=True
)
print("About to start")
server.start()

while not server.is_active:
    sleep(1)
    
print("SSH up")

server.stop()
while server.is_active:
    sleep(1)
print("Server stopped")
```

If I add:

```
server.daemon_forward_servers = True
server.daemon_transport = True
```

just before `server.start()` then it *does* work.

Note that it *does* work with the context manager:

```
import sshtunnel

with sshtunnel.open_tunnel(
        ssh_address_or_host='localhost',
        remote_bind_address=('localhost', 80),
        local_bind_address=('localhost', 8088),
        ssh_username='ec2-user',
        ssh_port=22,
        ssh_pkey='~/.ssh/epic_ssh_unit_test_key.private',
        ssh_config_file=None,
        allow_agent=False,
        set_keepalive=10,
        threaded=True
    ) as tunnel:

    print("SSH up")

print("Server stopped")
```

I think this is related to #169 , but hard to tell since there's no detail there.

**Comment by nickthorpe** [comment]
this is also causing issues for us in production. thanks for reporting @mdavis-xyz!

**Comment by pahaz** [comment]
Hmm. I assume that your problem may be related to https://github.com/pahaz/sshtunnel/pull/198/files#diff-11ccf18bb1618a9e107ff29db859eb1aaaa63d3aba94770f78de58deae467689R1412 
We have some similar discussions about it: https://github.com/pahaz/sshtunnel/issues/162 https://github.com/pahaz/sshtunnel/issues/173 

And it's maybe a problem for others users. 
Without that fix, we have an unclosed thread: https://github.com/pahaz/sshtunnel/issues/170

Currently, we are always waiting until closing all connections before the stop.
We assume that you know better how to stop your connections before stop the tunnel.
And you should close all your connections before call the `.stop()` function.

Is this a problem for you?

**Comment by pahaz** [comment]
You can also try this dirty hotfix:

```
x = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

server = SSHTunnelForwarder(
    ...,
    ssh_proxy=x,
)

```

It should revert changes in 0.2.0: https://github.com/pahaz/sshtunnel/pull/195

**Comment by mdavis-xyz** [comment]
> Could you please check is this related to 0.2.0? or 0.2.1?

I'm seeing this behavior in both.

In my case, I'm creating the tunnel, doing when HTTP request, then closing it. So I'm expecting the traffic has stopped by the time I call `.stop()`

**Comment by pahaz** [comment]
I think we can add something like `.stop(force=True)` which solves your case.

**Comment by pahaz** [comment]
You can try `stop(force=True)` in `0.2.2`.
Could you please test it?
Is this help you?

**Comment by dmulter** [comment]
I am using the context manager approach as I have for many years and the latest release is causing hangs on all my production deploys. Falling back to 0.1.5 until this major regression is resolved.

**Comment by pahaz** [comment]
> I am using the context manager approach as I have for many years and the latest release is causing hangs on all my production deploys. Falling back to 0.1.5 until this major regression is resolved. I see the failure under Python 3.6, but not 3.8.

Yes, it's possible. Thanks for the report. We have a plan to change the context behavior of 0.2.0+ here: https://github.com/pahaz/sshtunnel/pull/209 (on the next weekend)
This change should stabilize the behavior for different versions of Python.

We have a long story with some hack known as `daemon_forward_servers = True`, this hack is related to some old paramiko issues and different python version behavior. Currently, we don't need it anymore. And we have a plan to remove it.

It will be a 0.3.0 release.

**Comment by pahaz** [comment]
`sshtunnel==0.3.0` is released! if you still have any problem feel free to reopen the issue or open a new one.

**Comment by pahaz** [comment]
We released **0.4.0**. And changed the default threading daemon option for sshtunnel threads which should reduce hangs.
This issue will be locked. Feel free to open a new one if you have any problems.
